### DISCUSSION TITLE
Отбор любимых треков

### DISCUSSION THREAD

**Discussion by Maracumba** [Chimildic/goofy#113] [Помощь с алгоритмом] [answered]
Всем привет
Возникла такая идея в голове. Нужна помощь в коде.
Идея в том чтобы на основе трека(его особенностей) найти похожие по параметрам(+- отклонением) любимые треки. Что то вроде функции replaceWithSimilar.

**Comment by Chimildic** [comment, marked as answer]
Попробуем следующим образом. Вначале задать id трека, процент отклонения и ключи, чьи значения сравнивать. Чем больше ключей, тем меньше будет треков. У меня уже при трех ключах выходит один трек - исходный. 

```js
function findSimilarSavedTracks() {
  const TRACK_ID = '0evOmsnfWAqsGJlSTICgz7';
  const PERCENT = 5;
  const KEYS = ['danceability','valence'];

  let features = {};
  Object.entries(getCachedTracks([{ id: TRACK_ID }], { features: {} }).features[TRACK_ID]).forEach(item => {
    let key = item[0];
    let value = item[1];
    if (KEYS.includes(key) && !isNaN(value)){
      features[key] = { 
        value: value,
        min: value * (100 - PERCENT) / 100,
        max: value * (100 + PERCENT) / 100
      };
    }
  });

  let savedTracks = Source.getSavedTracks();
  Filter.rangeTracks(savedTracks, { features: features });
  Playlist.saveWithReplace({
    // id: '',
    name: 'Похоже на трек',
    tracks: savedTracks,
  });
}
```

**Reply by Bondian** [reply]
К слову, обнаружил для себя, что почти одинаковые по звучанию треки иногда имеют значительно отличающиеся "features". Местами отличия достигают более 20%
![Скриншот 08-08-2021 194646](https://user-images.githubusercontent.com/85184404/128636513-dac7f20c-1d9a-40f0-aad0-6fe4697cc391.png)
И наоборот - абсолютно разные по звучанию и стилистике треки имели идентичные "features".

**Comment by Chimildic** [comment]
Видоизмененный вариант, где задаются проценты под каждый ключ и источник это все треки из фонотеки
```js
function findSimilarSavedTracks() {
  const TRACK_ID = '0evOmsnfWAqsGJlSTICgz7';
  const RANGE = {
    energy: 4, 
    danceability: 8,
    valence: 15,
  }

  let features = {};
  let keys = Object.keys(RANGE);
  Object.entries(getCachedTracks([{ id: TRACK_ID }], { features: {} }).features[TRACK_ID]).forEach(item => {
    let key = item[0];
    let value = item[1];
    if (keys.includes(key) && !isNaN(value)){
      features[key] = { 
        value: value,
        min: value * (100 - RANGE[key]) / 100,
        max: value * (100 + RANGE[key]) / 100
      };
    }
  });

  let tracks = Source.getFollowedTracks({type: 'all'});
  Filter.rangeTracks(tracks, { features: features });
  Filter.dedupTracks(tracks);
  Playlist.saveWithReplace({
    // id: '',
    name: 'Похоже на трек',
    tracks: tracks,
  });
}
```

**Reply by Bondian** [reply]
Шикарный код! Никак не могу наиграться с различными источниками и исходными треками. Заметил, что большей похожестью звучания обладают треки, найденные с большим списком ключей, но с большим диапазоном отклонений. Например: `  const RANGE = {
    energy: 65,
    danceability: 65,
    valence: 65,
    acousticness: 65,
    anger: 65,
    mode: 65,
    key: 65,
    happiness: 65,
    sadness: 65,
    tempo: 65,
    time_signature: 65,
    // loudness: 65
  `}``
Да, кстати, по ключу loudness не находится даже исходный трек. Видимо, как-то сказывается отрицательное значение...

**Comment by Maracumba** [comment]
@Chimildic большое спасибо за оперативность. Попробовал скрипт, поизменял ключи, в итоге получается либо вообще мало, либо много непохожего на заданный трек.
А если использовать алгоритмы Спотифая через Source.craftTracks с заданными ключами и из результата оставлять только лайки?
я попробовал без указания ключей, из 100 треков остается 20 лайков.

**Reply by Chimildic** [reply]
Второй вариант не заточен под лайки. Без ключей количество треков не должно меняться. Если исключить фильтр дубликатов конечно.
Точной копии или большего количества не найдется. Слишком ограниченная выборка и количество сравниваемых ключей. Даже запрос рекомендаций у Spotify со всем набором ключей - не дает желаемых треков. 
`craftTracks` будет хуже, чем `replaceWithSimilar` в этом сценарии.

**Reply by Maracumba** [reply]
Получается все это дело не стоит свеч. Большое спасибо за отзывчивость. 
Попробую еще второй вариант
